File: Java/src/main/java/Solution101_200/Solution121.java
package Solution101_200;

/**
 * @author windows
 */
public class Solution121 {

    public int maxProfit(int[] prices) {
        int max = prices[prices.length - 1];
        int res = 0;
        for (int i = prices.length - 1; i >= 0; i--) {
            if (max - prices[i] > res){
               res = max - prices[i];
            }
            if (prices[i] > max) {
                max = prices[i];
            }
        }
        return res;
    }


    public static void main(String[] args) {
        Solution121 solution121 = new Solution121();
    }
}

File: C++/src/Solution101_200/Solution121.cpp
//
// Created by windows on 2023/3/1.
//

#include "Solution121.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

int Solution121::maxProfit(vector<int> &prices) {
    int maxProfit = 0;
    int maxValue = 0;
    for (int i = prices.size() - 1; i >= 0; i--){
        if (prices[i] > maxValue) {
            maxValue = prices[i];
        }else {
            if (maxValue - prices[i] > maxProfit){
                maxProfit = maxValue - prices[i];
            }
        }
    }
    return maxProfit;
}
